Basic Fragment Registration

KritR opened the pull request:
This is a basic implementation of fragments addressing #1 . It automatically adds
fragments to the process / global store and pulls from them as necessary
when mutations / queries are sent.

Things to be done:
- [x] Make fragment features independent module
- [x] Refactor out storage into separate module 
- [x] Add tests for fragment registration and parsing
- [x] Identify if "fragment" is within a string, and ignore parsing ( todo in code )

Some of these items do clutter neuron.ex. I'm wondering if this should reside in its own module?

Commit:
Added Basic Fragment Registration Implementation

This is a basic implementation of fragments. It automatically adds
fragments to the process / global store and pulls from them as necessary
when mutations / queries are sent.

Commit:
Fixed Lack of Type Conversions

There were some instances in which the proper types weren't being passed
around. There have been additional conversions added for this to be
fixed

uesteibar commented:
Hi @KritR!!

Thanks for taking the time for this, I really appreciate it ❤️ 

I would love to have some matters addressed before we actually dive into reviewing the implementation:

- As you pointed out, I think it would be better to extract this feature to its own module. I was thinking of `Neuron.Fragment.register`, but if you have a better idea feel free to go for it.
- Adding some tests with clear examples will also make it much easier to review.
- The [`Neuron.Config`](https://github.com/uesteibar/neuron/blob/master/lib/neuron/config.ex) module already implements some logic that is very similar to the one used in this PR to store/retrieve fragments. Would be ideal to either directly use `Neuron.Config` or extract the common parts to another module and use that in both to avoid some duplication.

After these points are addressed we can discuss implementation details and optimisations (such as the one you mentioned in the PR description).

Thanks again for the hard work!! Please feel free to reach out either on this PR or on my [personal email](mailto:uesteibar@gmail.com) to discuss any issues or proposals.

Commit:
Refactored Storage and Fragment into Modules

First of all, registering fragment was moved out of neuron.ex into its
own module ```Neuron.Fragment```. This kept things a little bit less
cluttered.

Storage functionality has also been moved out of Config and Fragment
into its own module. This ensures that code is not duplicated and that
whatever store is being used is abstracted away from the rest of the
code. Neuron.Store itself exposes a basic key value interface that
abstracts away the process / application data retrieval differences.

Commit:
Added Basic Tests for Fragments

Added a basic set of tests for fragments. Made changes to fragment.ex as
necessary to pass tests.
Removed failing / obsolete tests from config_test

Commit:
Added More Fragment Tests and Store Tests

Added all necessary tests for Neuron.Store. In addition, I updated the
doctest to pass.

Added a couple more fragment tests and made them pass as well. We've
fixed the problem of quoted fragments being parsed. Not sure what else
there is to do.

Commit:
Removed Obsolete TODO Marker

KritR commented:
Hey @uesteibar !

I've addressed all the matters listed above, in the todo list I created and the ones you mentioned as well. Let me know what you think!

uesteibar commented on lib/neuron/fragment.ex:
````diff
@@ -0,0 +1,63 @@
+  ## Example
+
+  ```elixir
+  Neuron.register("\"\"
````
Could we run this as a `DocTest`? This way we make sure we don't forget to update it if we change the API.

uesteibar commented on lib/neuron/fragment.ex:
```diff
@@ -0,0 +1,63 @@
+    {fragment_name, fragment}
+  end
+
+  def insert_into_query(query_string) do
```
Could we add `@doc false` here? Even though this function needs to be public, it's not intended to be used outside of the library.

uesteibar commented on lib/neuron/fragment.ex:
```diff
@@ -0,0 +1,63 @@
+
+    missing_fragments = fragments_to_add |> Enum.filter(&is_atom/1) |> Enum.map(&Atom.to_string/1)
+
+    if Enum.count(missing_fragments) > 0, do: raise("Fragments #{missing_fragments} not found")
```
We could use `Enum.any?/1` here

uesteibar commented on lib/neuron/store.ex:
```diff
@@ -0,0 +1,77 @@
+defmodule Neuron.Store do
+  @moduledoc """
+  This module provides a simple way to interface with the process and
+  application environmental variables, acting as a simple KV store.
```
I would add a warning here like:

```
WARNING: this module is not intended to be used outside of this library.
```

uesteibar reviewed (commented):
Left some tiny comments, other than that it LGTM! 🎉

uesteibar commented:
Hi @KritR! What's the status of this PR? Do you need help to move it forward? let me know :)

KritR commented:
Hey sorry about the delay. I’m on vacation for a little bit. I’ll make sure to review your reviews and push changes by the end of this week.

uesteibar commented:
@KritR sure! I just wanted to follow up on this. Enjoy your vacation!

Commit:
Added doc updates & simplified code

Changed example to doctest for Neuron.Fragment.register.
Added a doc false to insert_into_query.
Replaced Enum.count/1 > 0 with simply Enum.any?/1.
Added warning to store module to indicate not for use outside library.

uesteibar commented:
Great job here @KritR!!

Thank you ❤️ 🚀 🌔

uesteibar commented:
This is now published on [v0.7.0](https://hex.pm/packages/neuron/0.7.0)
